### PR TITLE
perf(messages): optimize performance

### LINKED ISSUE #545: optimize performance for messages component
Chats are taking about 10 seconds to render roughly 1000 messages, this can be improved.

Areas to fix:
- Messages are being mapped twice in the render when this can be changed to only once.
- Profile pictures are being fetched inside each individual msg component, fetch and cache conversation participant profile pictures inside the messages component and pass through to the msg components instead.

### PR DESCRIPTION
**Pull request by nathan-power**
### What this PR does 📖
Chats were taking about 10 seconds to render roughly 1000 messages. With these improvements it's only about 500ms to render 1000 messages, still not ideal but definitely an improvement.

Areas fixed:
- Messages were being mapped twice in the render but are now only mapped once.
- Profile pictures were being fetched inside each individual msg component. Conversation participant profile pictures are now fetched and cached inside the messages component and passed through to the msg components instead.

### Which issue(s) this PR fixes 🔨
- Resolve #545

### CODE REVIEW

File: src/components/main/compose/messages/mod.rs [hunk 1]
```diff
@@ -322,11 +322,23 @@ pub fn Messages(cx: Scope<Props>) -> Element {
...
     let senders: Vec<DID> = messages.iter().map(|msg| msg.sender()).collect();
+    let senders2 = senders.clone();
     // messages has already been reversed
     let idx_range = 0..messages.len();
     let next_sender = idx_range.clone().map(|idx| senders.get(idx + 1));
     let prev_sender = idx_range.map(|idx| if idx == 0 { None } else { senders.get(idx - 1) });
 
+    // get profile pictures for all senders in the conversation and cache them
+    let mut profile_pictures = HashMap::new();
+    for sender in senders2 {
```
Comment (dariusc93) [inline src/components/main/compose/messages/mod.rs]: Instead of cloning to `senders2` you could do 

```suggestion
    for sender in senders.iter() {
```

and clone `sender` where needed (although from a glance `senders` isnt needed either since one can grab the recipients from `Conversation` so that way youre not iterating over every single message just for say 2 or 3 recipieints but that could probably be in this pr or in a different pr)

File: src/components/main/compose/messages/mod.rs [hunk 2]
```diff
@@ -343,30 +355,36 @@ pub fn Messages(cx: Scope<Props>) -> Element {
...
                     let message_id = message.id();
                     let conversation_id = message.conversation_id();
                     let msg_sender = message.sender();
                     let is_remote = ident.did_key() != msg_sender;
                     let is_last = next_sender.map(|next_sender| *next_sender != msg_sender).unwrap_or(true);
                     let is_first = prev_sender.map(|prev_sender| *prev_sender != msg_sender).unwrap_or(true);
+                    let mut rg = rg.clone();
+
+                    let mut sender_picture = String::new();
+                    if let Some(pfp) = profile_pictures.get(&msg_sender) {
```
Comment (dariusc93) [inline src/components/main/compose/messages/mod.rs]: This could probably be changed to be like the following

```rust
let sender_picture = profile_pictures.get(&msg_sender).and_then(|pbp| pbp.clone()).unwrap_or_default();
```

**Review by dariusc93 — COMMENTED** [review]
Looks good so far. Left a couple of suggestions

**Review by phillsatellite — APPROVED** [review]

**Comment by phillsatellite** [comment]
Was good on my end when testing, put changes request for Darius's comments above

File: src/components/main/compose/messages/mod.rs (hunk 1 above)
Comment (nathan-power) [reply src/components/main/compose/messages/mod.rs]: Thanks, that's a much better way of doing it

Yeah the logic for senders could be handled a little better, I'll create a new issue for it

File: src/components/main/compose/messages/mod.rs (hunk 2 above)
Comment (nathan-power) [reply src/components/main/compose/messages/mod.rs]: Much cleaner, will change it now

File: src/components/main/compose/msg/mod.rs [hunk 3]
```diff
@@ -35,6 +36,7 @@ pub struct Props<'a> {
     first: bool,
     middle: bool,
     last: bool,
+    profile_picture: std::option::Option<std::string::String>,
```
Comment (dariusc93) [inline src/components/main/compose/msg/mod.rs:39]: Just as an fyi, you can just do `Option<String>` since rust would do `std::option::Option<std::string::String>` at compiletime but not a blocker here just a matter of keeping things clean.

**Review by dariusc93 — APPROVED** [review]

File: src/components/main/compose/msg/mod.rs (hunk 3 above)
Comment (nathan-power) [reply src/components/main/compose/msg/mod.rs:39]: That's good to know, thanks for the tip
